src/timesyncwait/timesyncwaitd.c

```diff
@@ -0,0 +1,124 @@
+        sd_event_source *event_source; /* non-null is the active io event source */
+} ClockState;
+
+static int io_handler(sd_event_source * s,
+                      int fd,
+                      uint32_t revents,
+                      void *userdata) {
+        ClockState *sp = userdata;
+        int rc = 0;
+        struct itimerspec its = {
+                .it_value.tv_sec = TIME_T_MAX,
+        };
+        struct timex tx = {0};
+
+        /* Release any event source and timer descriptor held from the last
+         * check, then set a timerfd to signal when the clock is set.  Do
+         * this before we check synchronized state to avoid a race
+         * condition. */
+        sp->event_source = sd_event_source_unref(sp->event_source);
+        if (0 <= sp->fd)
+                (void) close(sp->fd);
+        sp->fd = timerfd_create(CLOCK_REALTIME, TFD_NONBLOCK | TFD_CLOEXEC);
+        if (0 <= sp->fd)
+                rc = timerfd_settime(sp->fd, TFD_TIMER_ABSTIME | TFD_TIMER_CANCEL_ON_SET, &its, NULL);
+
+        if (0 == rc) {
+                /* Fetch the kernel time state.  Synchronization state is encoded
+                 * in the return value. */
+                sp->adjtime_state = rc = adjtimex(&tx);
+                if (0 <= rc) {
+                        fprintf(stderr, "adjtime state %d status %x time %u.%06u\n",
+                                sp->adjtime_state, tx.status,
+                                (unsigned)tx.time.tv_sec, (unsigned)tx.time.tv_usec / ((STA_NANO & tx.status) ? 1000 : 1));
+                }
+        }
+        if (TIME_ERROR == sp->adjtime_state)
+                /* Not synchronized.  Do a one-shot wait on the descriptor. */
+                rc = sd_event_add_io(sp->event, &sp->event_source, sp->fd,
+                                     EPOLLIN, io_handler, sp);
+        else if (0 <= sp->adjtime_state) {
+                /* Synchronized; we can exit. */
+                sd_event_exit(sp->event, 0);
+                return 0;
+        }
+        if (0 > rc)
+                sd_event_exit(sp->event, rc);
+        return rc;
+}
+
+int main(int argc,
+          char * argv[]) {
+        int rc;
+        ClockState state = {
+                .fd = -1,
+        };
+
+        rc = sd_event_default(&state.event);
+        if (0 <= rc) {
+                sigset_t ss;
+                if ((0 > sigemptyset(&ss))
+                    || (0 > sigaddset(&ss, SIGTERM))
+                    || (0 > sigaddset(&ss, SIGINT))
+                    || (0 > sigprocmask(SIG_BLOCK, &ss, NULL)))
+                        rc = -errno;
+        }
+        if ((0 <= rc)
+            && (0 <= ((rc = sd_event_add_signal(state.event, NULL, SIGTERM, NULL, NULL))))
+            && (0 <= ((rc = sd_event_add_signal(state.event, NULL, SIGINT, NULL, NULL))))
+            && (0 <= ((rc = sd_event_set_watchdog(state.event, true))))
+            && (0 <= ((rc = io_handler(0, -1, 0, &state))))
```

Comment:
please use "NULL" instead of "0" when initializing a pointer to null. i.e. the first argument here really should be NULL, not 0.

That said, I think it would be prettier if we wouldn't "fake" an event hanlder call here. Instead, let's make the io_handler() call call some common function with proper arguments, and let's then call that function here too.